src/cpp/src/greedy_decoding.cpp

```diff
@@ -0,0 +1,146 @@
+    results.tokens.resize(running_batch_size);
+    std::fill(results.scores.begin(), results.scores.end(), 0);
+    
+    int64_t kv_cache_len = 0;
+    if (is_chat_conversation && !is_cache_empty) {
+        OPENVINO_ASSERT(batch_size == 1, "continuation of generation is possible only for batch 1");
+        
+        // between subsequent runs attention_mask should not be modified
+        auto atten_mask_history = m_model_runner.get_tensor("attention_mask");
+        kv_cache_len = atten_mask_history.get_shape()[1];
+
+        size_t prompt_len = attention_mask.get_shape()[1];
+        ov::Tensor new_atten_mask =  ov::Tensor{ov::element::i64, {batch_size, kv_cache_len + prompt_len}};
+
+        std::copy(atten_mask_history.data<int64_t>(), atten_mask_history.data<int64_t>() + kv_cache_len,
+                  new_atten_mask.data<int64_t>());
+        std::copy(attention_mask.data<int64_t>(), attention_mask.data<int64_t>() + prompt_len,
+                  new_atten_mask.data<int64_t>() + kv_cache_len);
+
+        m_model_runner.set_tensor("attention_mask", new_atten_mask);
+    } else if (!is_cache_empty) {
+        OPENVINO_THROW("KV cache contains initial values but generate is run not in chat scenario. "
+                        "Initial KV cache can contain values only if start_chat() is called.");
+    } else {
+        m_model_runner.set_tensor("attention_mask", attention_mask);
+    }
+
+    if (position_ids_available) {
+        position_ids = ov::Tensor{ov::element::i64, input_ids.get_shape()};
+        utils::initialize_position_ids(position_ids, attention_mask, kv_cache_len);
+    }
+    
+    m_model_runner.set_tensor("input_ids", input_ids);
+    if (position_ids_available)
+        m_model_runner.set_tensor("position_ids", position_ids);
+
+    m_model_runner.get_tensor("beam_idx").set_shape({running_batch_size});
+    auto beam_data = m_model_runner.get_tensor("beam_idx").data<int32_t>();
+    std::iota(beam_data, beam_data + running_batch_size, 0);
+
+    m_model_runner.infer();
+    auto logits = m_model_runner.get_tensor("logits");
+    ov::Shape logits_shape = logits.get_shape();
+    size_t seq_len = logits_shape[1], vocab_size = logits_shape[2];
+    m_model_runner.get_tensor("input_ids").set_shape({running_batch_size, 1});
+
+    std::vector<int64_t> token_iter_results(running_batch_size);  // results of a single infer request
+    std::vector<int> eos_met(running_batch_size, 0);  // use int because can not use std::all_of with vector<bool>
+    for (size_t batch = 0; batch < running_batch_size; ++batch) {
+        auto out_token = utils::argmax(logits, batch);
+        results.tokens[batch].emplace_back(out_token);
+
+        token_iter_results[batch] = out_token;
+        eos_met[batch] = (out_token == generation_config.eos_token_id);
+        m_model_runner.get_tensor("input_ids").data<int64_t>()[batch] = out_token;
+    }
+    if (streamer && streamer->put(token_iter_results[0])) {
+        return results;
```

Comment:
@pavel-esir 
do we need to call `streamer->end()`?

Reply:
I didn't call `end()` here because of the following reason. Calling `end()` is needed for user to know that generation finished and `put()` will not be called anymore so he can flush the cache or do some stuff at the end if he has any. 

But here, we enter this this place only if user invoked stop / `return true` by him/herself, so he already know he want to stop, he can call `end()` by himself.

But if there are other reasons i missed, it's ok by me to call it here as well.